### PR TITLE
DOC: sparse: Add doc for migration to sparray

### PR DESCRIPTION
**Pull request by dschult**
This is a building ground for migration to sparray docs. Comments and ideas welcome!!

Where should this live?  `doc/dev`?

Is `rst` the right format?

### CODE REVIEW

**Comment by dschult** [comment]
This is a Hackmd style migration document (soon to be converted to rst).
@stefanv and @perimosocordiae Are there obvious categories of changes I am leaving out?

File: doc/source/dev/migration_to_sparray.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+The changeover from sparse matrices to sparse arrays involves changes
+in a few operations, changes in the output shape of some operations,
+and changes to the way some inpits are handled.
+
+### Obvious big picture differences:
+  - the constructor names *_matrix, e.g. csr_matrix, are changed to *_array.
+  - matrices are always 2D (rows x columns) even if there 1 row or column.
+    arrays can be 1D or 2D. Numpy scalars are used as 0D values where needed.  
+  - Operations change:
+    - matrices use operators * and @ for matrx multiplication, while arrays
```
Comment (stefanv) [inline doc/source/dev/migration_to_sparray.rst]:
```suggestion
    - matrices use operators * and @ for matrix multiplication, while arrays
```

File: doc/source/dev/migration_to_sparray.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      use * for elementwise multiplication and @ for matrix multiplication.
```
Comment (stefanv) [inline doc/source/dev/migration_to_sparray.rst]: What do matrices use for scalar multiply? I.e. what do I look for in existing code that needs to change?

Same for power below.

**Review by stefanv — COMMENTED** [review]
Nice, Dan, thank you!

In further shaping this, perhaps we could think about it in terms of user questions:

- In my code, what should change, from what to what?
- When writing new code, what should I watch out for, i.e. what is old vs new behavior.
- Are there any fundamental concepts / design choices that changed that I should keep in mind?

**Comment by perimosocordiae** [comment]
Let's start with Markdown first (to get things rendering nicely), and then if needed we can convert to RST later.

File: doc/source/dev/migration_to_sparray.rst [hunk 3]
```diff
@@ -0,0 +1,88 @@
+### Migration from spmatrix to sparray
+
+The changeover from sparse matrices to sparse arrays involves changes
+in a few operations, changes in the output shape of some operations,
+and changes to the way some inpits are handled.
```
Comment (perimosocordiae) [inline doc/source/dev/migration_to_sparray.rst]: inpits -> inputs

File: doc/source/dev/migration_to_sparray.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
+### Migration from spmatrix to sparray
+
+The changeover from sparse matrices to sparse arrays involves changes
```
Comment (perimosocordiae) [inline doc/source/dev/migration_to_sparray.rst]: Let's start by saying that this mirrors the change from np.matrix to np.array, which hopefully we can find some useful pages on the web for. I wasn't able to find a canonical "migration from np.matrix to np.array" page though.

File: doc/source/dev/migration_to_sparray.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+# Any use of these should be examined and return values treated as sparrays.
+def kron(A, B, format=None):
+def kronsum(A, B, format=None):
+def hstack(blocks, format=None, dtype=None):
+def vstack(blocks, format=None, dtype=None):
+def block_diag(mats, format=None, dtype=None):
+def tril(A, k=0, format=None):
+def triu(A, k=0, format=None):
+
+# Names that should be rewritten to return sparse arrays
```
Comment (perimosocordiae) [inline doc/source/dev/migration_to_sparray.rst]: A table with these would be very nice.

File: doc/source/dev/migration_to_sparray.rst (hunk 4 above)
Comment (dschult) [reply doc/source/dev/migration_to_sparray.rst]: I believe there never really was a migration from np.matrix to np.array.
Stefan and I were not able to find such previous documentation.

Still, I've added text to this point at the top.

**Pull request closed without merging**